lib/new_relic/traced_thread.rb

```diff
@@ -0,0 +1,27 @@
+# encoding: utf-8
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-ruby-agent/blob/main/LICENSE for complete details.
+# frozen_string_literal: true
+
+module NewRelic
+  class TracedThread < Thread
+    def initialize(*args, &block)
+      # we should add metrics
+      traced_block = create_traced_block(*args, block)
+      super(*args, &traced_block)
+    end
+
+    def create_traced_block(*args, block)
+      return block if NewRelic::Agent.config[:'instrumentation.thread.tracing'] # if this is on, don't double trace
+
+      instrumentation = ::Thread.current[:newrelic_tracer_state]
+      Proc.new do |*args|
+        ::Thread.current[:newrelic_tracer_state] = instrumentation
+        segment = NewRelic::Agent::Tracer.start_segment(name: "Thread#{::Thread.current.object_id}")
```

Comment:
possible future feature idea - allow customers to set the name of the thread

Reply:
hmm interesting idea. right now all args passed in are passed straight to the thread, but maybe we could peek at them to just see if something specific is present for the name.
also, yeah i wasn't sure what to really call the segment, i'm definitely open to talking about it and seeing what might be a better name also.